File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    def bfs(x, y):
        q = deque()
        q.append([x, y, 1])

        while q:
            x, y, distance = q.popleft()

            if x == len(maps) - 1 and y == len(maps[0]) - 1:
                return distance

            if maps[x][y] == 0: continue
            maps[x][y] = 0

            if x + 1 < len(maps):
                q.append((x + 1, y, distance + 1))
            if y + 1 < len(maps[0]):
                q.append((x, y + 1, distance + 1))
            if x - 1 >= 0:
                q.append((x - 1, y, distance + 1))
            if y - 1 >= 0:
                q.append((x, y - 1, distance + 1))

        return -1
    return bfs(0, 0)
    
    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    let minCount = 0
    
    function bfs(x, y) {
        const queue = []
        queue.push([x, y, 1])
        
        while (queue?.length > 0) {
            const [x, y, distance] = queue.shift()
            
            if (x === maps.length - 1 && y === maps[0].length - 1) {
                return distance
            }
            
            if (maps[x][y] === 0) {
                continue
            }
            
            maps[x][y] = 0
            
            if (x + 1 < maps.length) {
                queue.push([x + 1, y, distance + 1])
            }
            if (y + 1 < maps[0].length) {
                queue.push([x, y + 1, distance + 1])
            }
            if (x - 1 >= 0) {
                queue.push([x - 1, y, distance + 1])
            }
            if (y - 1 >= 0) {
                queue.push([x, y - 1, distance + 1])
            }
        }
        
        return -1
    }
    
    return bfs(0, 0)

}